Fix single quiz selection flow

rtibbles opened the pull request:
## Summary
* Updates routing for quiz creation to use sectionIndex and make the URL the source of truth for the active section
* Updates question titles to be blank when not set by user for future updates when we allow them to be editable
* Updates question title display to account for this.
* Fixes a previously unreported bug where the fetch for resources was not filtering properly
* Updates strings for the SectionEditor component
Before:
![Screenshot from 2024-06-11 17-15-55](https://github.com/learningequality/kolibri/assets/1680573/61778623-fd86-46b6-b23f-57a2636eaddb)
After:
![Screenshot from 2024-06-11 17-27-40](https://github.com/learningequality/kolibri/assets/1680573/4c8ca299-59a6-4c8f-a641-51fdaacd7ef7)
* Updates the select quiz workflow to use the new components and state, with some minor tweaks for the quiz creation flow for this:
  * Separates out resource_pool updating to a separate method so that we can directly set values without triggering randomized question setting
  * Conditionalizes the resource selection modal to behave differently depending on whether we are selecting a quiz
* Cleans up all unused components and vuex state

## References
Fixes #12247

## Reviewer guidance
Ensure you have practice quizzes imported.
Choose the 'select a quiz' option from the create quiz dropdown.
Ensure you can select a quiz and then create a quiz from there.
Ensure that this has not induced any regressions in the main quiz creation experience.

----

## Testing checklist

- [ ] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Critical and brittle code paths are covered by unit tests


## PR process

- [ ] PR has the correct target branch and milestone
- [ ] PR has 'needs review' or 'work-in-progress' label
- [ ] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

## Reviewer checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

rtibbles committed:
Allow any data type to define a functional default.

rtibbles committed:
Update channel API to add filtering by quizzes and make consistent with contentnode endpoint.

rtibbles committed:
Update question titles to be blank if not set by user.
Update exam utils spec to properly assert conversions being tested against V3.

rtibbles committed:
Fix bug with resource fetching parameters.

rtibbles committed:
Clean up unused components and state.
Update QuestionListPreview for just read only usage.
Update display of questions.

pcenov commented:
Hi @rtibbles,

1. I am always getting a server error when I select a practice quiz (or just create a quiz from scratch), save it and then attempt to start it:

https://github.com/learningequality/kolibri/assets/79847249/0470f2ee-c10c-4bc1-88ef-32777fed927f

2. When I am adding a section, or replacing questions and I click 'Save changes' the modal remains open and I have to click multiple times the back arrow to exit it:

https://github.com/learningequality/kolibri/assets/79847249/215635d9-9f34-440f-9d2e-bd744f5011a2

3. When I create a brand new quiz and I click the Save button there's a brief flash of the validation message for the quiz title, and then if I click the Save button again it's no longer possible to save the quiz with the same title:

https://github.com/learningequality/kolibri/assets/79847249/50b4eb01-194f-49a2-bf52-c99ebd33496f

4. I can't delete a single or a few of the questions in the section, getting a `TypeError: Cannot update both `questions` and `question_count` at the same time; use one or the other.` Deleting all of the questions is working though.

https://github.com/learningequality/kolibri/assets/79847249/439452d7-6ec3-48a2-bfdd-f143f60ca9f9

Logs: [logs.zip](https://github.com/user-attachments/files/15806664/logs.zip)

rtibbles commented:
Thanks @pcenov!

nucleogenesis commented:
@rtibbles I think Peter's #3 can be covered in https://github.com/learningequality/kolibri/issues/12262

rtibbles committed:
Remove previous route tracking in resource selection.
Just return to root quiz page on save.

rtibbles committed:
Fix all references to root quiz creation route to include sectionIndex.

rtibbles committed:
Remove question_count from sections.

rtibbles committed:
Update remaining references to question id to item.

rtibbles commented:
I think I have addressed all issues bar 3 - which I think @nucleogenesis is tackling elsewhere.

I think this also addresses [#12246](https://github.com/learningequality/kolibri/issues/12246).

It also updates the resource selection flow to be about adding new questions specifically, with each 'update resources' allowing you to choose how many questions you are selecting and choosing resources to fill those questions.

I implemented this after discussion with @marcellamaki and @nucleogenesis about the counterintuitiveness of selecting resources first and only _then_ deciding the number of questions. Interest in thoughts and feedback, although I think a lot of the strings could definitely do with being reworded.

pcenov commented:
Hi @rtibbles, I've had a look at the latest changes here and:

1. When I select a quiz, then none of the questions are getting displayed in the section. There's a `TypeError: Cannot read properties of undefined (reading 'title')` error in the console. It's also not possible to add the questions manually:

https://github.com/learningequality/kolibri/assets/79847249/960f360c-a8fd-451c-8e08-d03baa5b6edb

2. Similarly when I create a new quiz, add a new section and attempt to add questions I am getting a "TypeError: Cannot read properties of undefined (reading 'ancestors')" error in the console and I can neither add more questions or close the modal by clicking outside:

https://github.com/learningequality/kolibri/assets/79847249/301ef7ca-d47d-4af2-bb6b-7792607b3e7d

rtibbles commented:
Irritating, I had both those errors in development, but thought I had addressed them - sorry for the wild goose chase.

nucleogenesis reviewed (changes requested):
**Manual testing notes**

#### Resource selection # questions selected, resources previously selected


https://github.com/learningequality/kolibri/assets/6356129/699f6a0a-be12-4ef7-93d2-fac33c1e35d9



#### Number of questions changes not save-able unless you change your resource selection (and the selectAll is visible at the end when it shouldn't be)


https://github.com/learningequality/kolibri/assets/6356129/849b8e14-cfe5-4d6c-aa76-c03e8a1ddc1d



#### Adding resources adds a number of questions in excess of the defined `question_count` on save, but does not impact the value in the form input


https://github.com/learningequality/kolibri/assets/6356129/69c52382-b38c-4a94-a5c4-5ad86104d2fa

rtibbles committed:
Update validation and messaging to better guide users to add questions.

rtibbles committed:
Handle practice quizzes longer than "MAX_QUESTIONS" by spreading them across multiple sections.

rtibbles commented:
@pcenov I replicated and think I have fixed both of the issues you experienced.

@nucleogenesis I updated copy to make it clear that all of the issues you raised are now intended behaviour (we have also talked about this extensively in person and on slack).

rtibbles committed:
Enforce section questions max length via API serializer.

pcenov commented:
Hi @rtibbles - I confirm that both issues are fixed now. I'll just mention here some other issues which I noticed while regression testing but maybe it's better to have everything merged and then to test everything again:

1. Missing snackbar confirmation message when saving a quiz:

https://github.com/learningequality/kolibri/assets/79847249/0421fb58-2730-49ac-831a-04cb1e547a61

2. When I want to replace questions if I select more than the needed number of questions and then try to correct the mistake, the Replace button remains disabled and the number of questions in "Select N fewer questions to continue" is incorrect:

https://github.com/learningequality/kolibri/assets/79847249/86db03bc-01b9-4c02-beac-07f80fc6d111

3. When adding questions, there needs to be more padding under the text "Each section may only contain up to 50 questions, this section has 0":

![2024-06-14_13-30-46](https://github.com/learningequality/kolibri/assets/79847249/70bf89f8-17f2-4111-850d-1154f8bee569)

4. It's still not possible to exit the "Add questions" modal by clicking outside of it (as is the case with the "Options" modal", so if I have been exploring a channel I have to repeatedly click the back arrow and confirm the "Are you sure you want to leave this page?" prompt before being able to actually exit the modal:

https://github.com/learningequality/kolibri/assets/79847249/36cf54e5-cfa1-473c-99c0-ea25dadc9a93

rtibbles commented:
Yeah, I can fix 3 here, as that text has been added within this PR, but the other issues are probably best dealt with elsewhere!

We are planning on holding off merging this PR until Monday, so that we can test out the original workflow for managing questions by adding and removing resources vs the newer workflow of just incrementally adding questions from resources that I have implemented here.

Glad to know that it is generally working though!

rtibbles committed:
Add margin top to topic warning.
Make string parameterized in case of changes.

rtibbles commented:
Have updated the margin top on the topic warning
![Screenshot from 2024-06-14 07-46-17](https://github.com/learningequality/kolibri/assets/1680573/2f2f4ffa-0cbc-4ce3-8e1b-9d001a4c78ed)

nucleogenesis commented on kolibri/plugins/coach/assets/src/views/plan/CreateExamPage/ResourceSelection.vue:
```diff
@@ -483,7 +619,12 @@
         resetWorkingResourcePool,
         contentPresentInWorkingResourcePool,
-        //contentList,
+        questionCount,
```
`contentPresentInWorkingResourcePool` (named on the line above) is failing for me when I go through to the QA Channel -> Exercises

I have a KA Perseus topic which has no children so when the card for that KA Perseus topic is passed to `contentPresentInWRP` it has no children, which chokes up the function.

nucleogenesis replied:
Might just work well to filter out empty topics to begin with - which I thought we'd been doing 🤔

nucleogenesis commented on kolibri/plugins/coach/assets/src/views/plan/LessonResourceSelectionPage/ContentCardList.vue:
```diff
@@ -16,14 +16,23 @@
           @change="handleCheckboxChange(content, $event)"
         />
+        <KRadioButton
+          v-else-if="contentHasCheckbox(content) && showRadioButtons"
```
Non-blocking, minor bug:

When I select the quiz and hit the confirm button, the KRadioButton disappears briefly before redirecting. It is _very_ pronounced on "Fast 3G" throttling.

nucleogenesis replied:
I think maybe "just use this on Fast 3G for a bit and fix what you see" could be a solid theme for a cohack, some things I noticed before I couldn't handle it anymore:

_Note, I have my cache disabled as well, so all typically cached API calls & resources would likely be much faster, but this is worth polishing_

- Click "Add section" -- the section tab is added... then nothing for a few seconds... then the slideout appears w/ the section settings
- Click "Add questions" -- basically the same, nothing happens - the page doesn't start spinning... it just sits there until the new content is displayed
- Basically a lot of transitions are met with a blank stare and sudden change rather than a spinner or some other indication of "something is happening just wait a sec"

Again -- not blocking, just noticed during testing here and can put into follow-up today

nucleogenesis commented on kolibri/plugins/coach/assets/test/selectRandomExamQuestions.spec.js:
```diff
@@ -1,13 +1,25 @@
-  ['A:A1', 'A:A2', 'A:A3'],
-  ['B:B1', 'B:B2', 'B:B3', 'B:B4', 'B:B5', 'B:B6', 'B:B7', 'B:B8'],
-  ['C:C1', 'C:C2', 'C:C3', 'C:C4', 'C:C5', 'C:C6', 'C:C7', 'C:C8', 'C:C9'],
+const EXERCISES = [
+  {
+    id: 'A',
+    title: 'title_x',
+    assessmentmetadata: { assessment_item_ids: ['A1', 'A2', 'A3'] },
+  },
+  {
+    id: 'B',
+    title: 'title_y',
+    assessmentmetadata: { assessment_item_ids: ['B1', 'B2', 'B3', 'B4', 'B5', 'B6', 'B7', 'B8'] },
+  },
+  {
+    id: 'C',
+    title: 'title_z',
+    assessmentmetadata: {
+      assessment_item_ids: ['C1', 'C2', 'C3', 'C4', 'C5', 'C6', 'C7', 'C8', 'C9'],
+    },
+  },
```
So much prettier <3

nucleogenesis commented on kolibri/plugins/coach/assets/src/utils/selectQuestions.js:
```diff
@@ -9,35 +8,55 @@ const logging = logger.getLogger(__filename);
 
 const getTotalOfQuestions = sumBy(qArray => qArray.length);
 
+export function exerciseToQuestionArray(exercise) {
+  return exercise.assessmentmetadata.assessment_item_ids.map((question_id, i) => {
+    return {
+      exercise_id: exercise.id,
+      question_id,
+      counter_in_exercise: i + 1,
+      // In the V3 schema, the title is user editable, and no longer
+      // simply the title of the exercise the question came from.
+      // We set it to blank to indicate that no user generated title
+      // has been created.
+      title: '',
+      item: `${exercise.id}:${question_id}`,
+    };
+  });
+}
```
Love this

nucleogenesis reviewed (changes requested):
Only one blocking error and some comments - my review isn't complete yet but I have to come back to finish later on.

rtibbles committed:
Remove use of section_id for routing of sections. Prefer sectionIndex.

rtibbles committed:
Update strings for section editor.

rtibbles committed:
Update select quiz workflow to use new components and composable.

rtibbles committed:
Update resource selection to add a specific number of resources each time.
Tweak question replacement.

rtibbles committed:
Clean up unused contentnode api endpoint.

rtibbles committed:
Ensure that query params are validated before being used to query.

rtibbles committed:
Move MAX_QUESTIONS to core constants. Rename for clarity.
Update v3 conversion logic to constrain section length.

rtibbles committed:
Don't show num assessments info for practice quiz selection.
Ensure we don't show any topics without children, even if the unfiltered num assessments says otherwise.

nucleogenesis reviewed (commented):
Non-blocking issue:

- Select a practice quiz with only a few questions
- Select those questions and then click to replace them
- You shouldn't have enough replacements, now if you click "Add resources" on that model, you open the Resource Selection

The issue is that the resource side panel opens to the Practice Quiz resource which I've already added instead of the channels list

nucleogenesis reviewed (changes requested):
Save followed by immediate Save & Close isn't working here:


https://github.com/learningequality/kolibri/assets/6356129/2fc246d3-bf23-4b4b-bb94-e33ddfcf6eb2

rtibbles committed:
Ensure we reset quizHasChanged on save.

rtibbles committed:
Only reuse necessary params from the route.
Update strings for replacements modal to reflect updated workflow.

rtibbles commented:
I think both of the things you observed here should now be fixed @nucleogenesis - I have also tweaked the warning modal shown when there are not enough replacements to better represent the current state in this PR.

radinamatic commented:
I'm struggling a bit around wrapping my brain around all the changes here... 😅 

Some re-wording suggestions:

1. Questions here are _past_ selected (since we use that term for states where user actively uses checkboxes to _select_), so maybe this string should be straight-up: **Number of questions: 5**

   <img width="1200" alt="2024-06-17_02-33-41" src="https://github.com/learningequality/kolibri/assets/1457929/98cee294-a80a-4942-9a58-a6d69cc7a0e4">


2. I suggest we simplify this whole area (marked in orange in the screenshot below) -   cc @tomiwaoLE  

   ---
   ⬅️ **Add questions to <section_name>** (modal title - H1)

   Current number of questions: **5**
   Number of questions to add: (input field with the counter, default value 0)
   
   **Select exercise resources to add questions**
   
   <`breadcrumbs`>
   <`search field`>
   ...

   ---

   <img width="1200" alt="2024-06-17_02-26-57" src="https://github.com/learningequality/kolibri/assets/1457929/4e70b2c7-467f-4e92-8123-2eff31b965bb">

3. Marked in red above is the overlap between question description and the string **`N questions unused`**. I would also maybe suggest the change to **`N unused questions available`**.

4. If we do go with my suggestion above it might not be visible, but even after I installed the latest asset today, there's still missing margin/padding above the search field:

Firefox | Chrome
-- | --
<img width="1200" alt="firefox-missing-margin" src="https://github.com/learningequality/kolibri/assets/1457929/cb45355c-db7f-415c-b401-1c6a15932af8"> | <img width="1200" alt="chrome-missing-margin" src="https://github.com/learningequality/kolibri/assets/1457929/abac3151-5867-45a7-9394-1e49dd2957c9">


5. For some reason I'm getting the network error when I try to play the videos above by @pcenov and @nucleogenesis , so this might have already been mentioned, but something is borked in the counter of questions to replace:


https://github.com/learningequality/kolibri/assets/1457929/c6459303-f4aa-4128-9c2c-8eb749a05175

nucleogenesis reviewed (approved):
This looks good - let's merge and the issues noted in the bug bash can be replicated and made into issues today
